src/terrain/clients/iplant_groups.clj

```diff
@@ -57,3 +60,49 @@
     (when-not (#{200 404} status)
       (throw (Exception. (str "iplant-groups service returned status " status))))
     {:subjects (:subjects (:body res))}))
+
+(defn- get-client []
+  (c/new-cyverse-groups-client (config/ipg-base) (config/environment-name)))
+
+(defn- get-optional-folder [client user name]
+  (try+
+   (c/get-folder client user name)
+   (catch [:status 404] _ nil)))
+
+(defn- create-user-folder [client user name]
+  (let [folder (c/add-folder client (config/grouper-user) name "")]
+    (c/grant-folder-privilege client (config/grouper-user) name user :stem)
+    folder))
+
+(defn- get-or-add-folder [client user name]
+  (or (get-optional-folder client user name)
+      (create-user-folder client user name)))
+
+(defn- get-collaborator-list-folder [user client]
+  (let [name (c/build-folder-name client (format "users:%s:collaborator-lists" user))]
+    (get-or-add-folder client user (string/replace name  #":[^:]+$" ""))
+    (get-or-add-folder client user name)))
+
```

Comment:
I think we also still want to create the `collab-folder` if it doesn't exist, so the current `get-or-add-folder` covers both the user folder and the collab folder.

I'm used to the `or` short-circuiting expression, but I can see how it could be trickier to read, and when I first reviewed this PR, I also almost suggested this change:

```
(defn- get-or-add-folder [client user name]
  (try+
    (c/get-folder client user name)
    (catch [:status 404] _
      (create-user-folder client user name))))
```